crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -546,8 +542,50 @@ async fn execute_transactions(
 
     authority_state.transaction_manager.enqueue(txns).await?;
 
-    authority_state
+    let effects = authority_state
         .database
         .notify_read_effects(tx_digests)
-        .await
+        .await?;
+
+    assert_eq!(
+        effects.len(),
+        num_digests,
+        "Missing txn effects from transaction manager"
+    );
+    Ok(effects)
+}
+
+fn get_transactions(
+    digests: &Vec<TransactionDigest>,
+    authority_state: Arc<AuthorityState>,
+) -> SuiResult<Vec<VerifiedCertificate>> {
+    let synced_txns = authority_state
+        .database
+        .perpetual_tables
+        .synced_transactions
+        .multi_get(digests)?
+        .into_iter()
+        .flatten()
+        .map(|tx| tx.into());
+
+    let executed_txns = authority_state
+        .database
+        .perpetual_tables
+        .certificates
+        .multi_get(digests)?
+        .into_iter()
+        .flatten()
+        .map(|tx| tx.into());
+
+    let pending_txns = authority_state
+        .database
+        .epoch_store()
+        .multi_get_pending_certificate(digests)?
+        .into_iter()
+        .flatten();
```

Comment:
this redundantly fetches every cert up to 3 times. Shouldn't we only try to load digests that weren't found in the previous source?